sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,376 @@
+      return new Item<>(namespace, key, type, value, null, null);
+    }
+
+    private Item(String namespace, String key, Type type, T value, String url, String label) {
+      this.ns = namespace;
+      this.key = key;
+      this.type = type;
+      this.value = value;
+      this.url = url;
+      this.label = label;
+    }
+
+    String getNamespace() {
+      return ns;
+    }
+
+    Type getType() {
+      return type;
+    }
+
+    T getValue() {
+      return value;
+    }
+
+    String getLabel() {
+      return label;
+    }
+
+    String getUrl() {
+      return url;
+    }
+
+    String getKey() {
+      return key;
+    }
+
+    /**
+     * Retrieve a serialized version of the display item value suitable for use as JSON.
+     */
+    String getValueString() {
+      return type.serializeJsonString(value);
+    }
+
+    private Item<T> addLabel(String label) {
+      return new Item<>(this.ns, this.key, this.type, this.value, this.url, label);
+    }
+
+    private Item<T> addUrl(String url) {
+      return new Item<>(this.ns, this.key, this.type, this.value, url, this.label);
+    }
+  }
+
+  /**
+   * Unique identifier for a display metadata item within a {@link PTransform} instance.
+   * Identifiers are composed of the key the are registered with and a namespace generated from
+   * the class of the component which registered the item.
+   *
+   * <p>Display metadata registered with the same key from different components will have different
+   * namespaces and will thus both be represented in the composed {@link DisplayData}. If a
+   * single component registers multiple metadata items with the same key, only the most recent
+   * item will be retained; previous versions are discarded.
+   */
+  @AutoValue
+  abstract static class Identifier {
+    static Identifier of(String namespace, String key) {
+      return new AutoValue_DisplayData_Identifier(namespace, key);
+    }
+
+    abstract String getNamespace();
+
+    abstract String getKey();
+  }
+
+  /**
+   * Display metadata type.
+   */
+  static enum Type {
+    STRING {
+      @Override
+      String serializeJsonString(Object value) {
+        return (String) value;
+      }
+    },
+    INTEGER {
+      @Override
+      String serializeJsonString(Object value) {
+        return Long.toString((long) value);
+      }
+    },
+    FLOAT {
+      @Override
+      String serializeJsonString(Object value) {
+        return Double.toString((Double) value);
+      }
+    },
+    TIMESTAMP() {
+      @Override
+      String serializeJsonString(Object value) {
+        return Long.toString(((Instant) value).getMillis());
+      }
+    },
+    DURATION {
+      @Override
+      String serializeJsonString(Object value) {
+        return Long.toString(((Duration) value).getMillis());
+      }
+    },
+    JAVA_CLASS {
+      @Override
+      String serializeJsonString(Object value) {
+        return ((Class<?>) value).getName();
```

Comment:
I believe the proposal suggested encoding the getSimpleName() and the getName() separately, so that rendering systems don't need to parse the name and duplicate that logic.